progressive_rollout_demo/lib/main.dart

```diff
@@ -0,0 +1,150 @@
+import 'dart:math';
+
+import 'package:cloud_firestore/cloud_firestore.dart';
+import 'package:firebase_core/firebase_core.dart';
+import 'package:flutter/material.dart';
+import 'package:package_info_plus/package_info_plus.dart';
+import 'package:shared_preferences/shared_preferences.dart';
+import 'package:shorebird_code_push/shorebird_code_push.dart';
+
+const groupKey = 'groupNumber';
+const firestoreCollectionName = 'rollouts';
+const firebaseDocumentName = 'rollouts';
+
+/// Assigns a random group number (1-100) to the user if one has not already
+/// been assigned.
+Future<void> assignGroupNumberIfNeeded() async {
+  final prefs = await SharedPreferences.getInstance();
+  if (!prefs.containsKey(groupKey)) {
+    final groupNumber = Random().nextInt(100) + 1;
+    await prefs.setInt(groupKey, groupNumber);
+  }
+}
+
+Future<void> main() async {
+  WidgetsFlutterBinding.ensureInitialized();
+
+  await Firebase.initializeApp();
+  await assignGroupNumberIfNeeded();
+
+  runApp(const MyApp());
+}
+
+class MyApp extends StatelessWidget {
+  const MyApp({super.key});
+
+  @override
+  Widget build(BuildContext context) {
+    return MaterialApp(
+      theme: ThemeData(
+        colorScheme: ColorScheme.fromSeed(seedColor: Colors.deepPurple),
+        useMaterial3: true,
+      ),
+      home: const MyHomePage(),
+    );
+  }
+}
+
+class MyHomePage extends StatefulWidget {
+  const MyHomePage({super.key});
+
+  @override
+  State<MyHomePage> createState() => _MyHomePageState();
+}
+
+class _MyHomePageState extends State<MyHomePage> {
+  final _updater = ShorebirdUpdater();
+
+  int? _groupNumber;
+
+  /// Returns the current release version (e.g., 1.0.0+1).
+  Future<String> _releaseVersion() async {
+    final packageInfo = await PackageInfo.fromPlatform();
+    return '${packageInfo.version}+${packageInfo.buildNumber}';
+  }
+
+  /// Fetches the rollout percentage for the current release version
+  /// (e.g., 1.0.0+1) from Firestore.
+  ///
+  /// You can view the Firestore data in the console at:
+  /// https://console.firebase.google.com/u/0/project/progressive-rollout-demo/firestore/databases/-default-/data/~2Frollouts~2Frollouts
```

Comment:
Was it your intention for firestore to be publicly editable?

Reply:
Nope, I don't think it is—can you access that without being logged into your shorebird account?

Reply:
![Screenshot 2024-11-13 at 3 57 04 PM](https://github.com/user-attachments/assets/e509d423-724f-4219-aa40-5f7bfefa14ef)

I see this when trying to view the page logged in to my gmail account

Reply:
ah okay so is the intention that we'd remove the link since it's not publicly available?

Reply:
Added a note that this is only accessible to shorebird employees. I want to leave breadcrumbs for us here when we want to add to/update this guide.